File: frontend-v2/src/lib/nightShift.ts
/**
 * Night shift — frontend model (ROADMAP E2).
 *
 * The operator marks tasks "run tonight" (a harness × runtime pair, local by
 * default); mc-worker starts them one after another as heads inside the
 * window from Settings and sends one morning report. Types for
 * `/api/v1/night-shift*` and the pure rules the UI follows.
 *
 *   - A pair may be marked when it could start now OR is only held back for
 *     today (`engine_not_ready`, `box_busy`) — by tonight the model can be up
 *     and the box free. Every other block stays a block.
 *   - The pre-selected pair is the head launcher's local default; a cloud
 *     pair is never pre-selected.
 *   - Texts never come from the backend: state and reason CODES map to keys
 *     in the `nightShift` i18n namespace.
 *   - MC is the operator's channel: the morning report and blocked notices
 *     show on Home ("Last night" card). Slack / Telegram get a copy only
 *     when `send_to_channels` is on (default off).
 */

import { chooseInitialPair, parseHeadError, type HeadPair, type HeadPairsResponse, type HeadRun } from "./heads";

export interface NightWindow {
  night: string;
  starts_at: string;
  ends_at: string;
}

export interface NightConfig {
  enabled: boolean;
  start: string;
  end: string;
  timezone: string;
  cloud_share: number;
  /** also send the report / notices to Slack or Telegram (default off) */
  send_to_channels: boolean;
  /** report channels configured right now ("telegram", "slack") */
  channels: string[];
  /** now inside the window */
  active: boolean;
  /** the current window, or the next one */
  window: NightWindow;
}

export type NightConfigUpdate = Partial<
  Pick<NightConfig, "enabled" | "start" | "end" | "timezone" | "cloud_share" | "send_to_channels">
>;

export type NightEntryState = "queued" | "waiting" | "skipped" | "started";

export interface NightEntry {
  task_id: string;
  title: string;
  task_status: string;
  harness: string;
  runtime_slug: string;
  locality: "local" | "cloud";
  marked_at: string | null;
  night: string | null;
  state: NightEntryState;
  reason: string | null;
  run: HeadRun | null;
}

export type NightCategory = "passed" | "failed" | "needs_you" | "blocked" | "running";

export interface NightReportEntry {
  task_id: string;
  title: string;
  started: boolean;
  state?: string | null;
  reason?: string | null;
  blocked?: string | null;
  pr_url?: string | null;
  category: NightCategory;
}

export interface NightReport {
  night: string;
  sent_at: string;
  delivered: boolean;
  /** sent · undelivered (channels on, sending failed) · stored (channels off: MC only) */
  state?: "sent" | "undelivered" | "stored" | "sending";
  entries: NightReportEntry[];
}

export interface NightTonight {
  config: NightConfig;
  entries: NightEntry[];
  last_report: NightReport | null;
}

// ── Last night (Home card) ───────────────────────────────────────────────────

export interface LastNightEntry extends NightReportEntry {
  run_id?: string | null;
  silent_s?: number | null;
  harness?: string | null;
  runtime_slug?: string | null;
  /** the live run — a question may be answered by now */
  run: HeadRun | null;
}

export interface LastNightReport {
  night: string;
  sent_at: string | null;
  /** also delivered to Slack / Telegram */
  delivered: boolean;
  dismissed: boolean;
  entries: LastNightEntry[];
}

/** A night head that is blocked right now (shown during the night). */
export interface NightNotice {
  task_id: string;
  title: string;
  kind: "needs_you" | "silent";
  silent_s: number | null;
  run: HeadRun;
}

export interface LastNight {
  report: LastNightReport | null;
  notices: NightNotice[];
}

/** Card order: what needs the operator first, what went fine last. */
const CARD_ORDER: NightCategory[] = ["needs_you", "blocked", "failed", "running", "passed"];

/** Report rows in card order (stable within a category). */
export function lastNightRows(report: Pick<LastNightReport, "entries"> | null | undefined): LastNightEntry[] {
  const rank = (c: NightCategory) => {
    const i = CARD_ORDER.indexOf(c);
    return i < 0 ? CARD_ORDER.length : i;
  };
  return [...(report?.entries ?? [])]
    .map((e, i) => ({ e, i }))
    .sort((a, b) => rank(a.e.category) - rank(b.e.category) || a.i - b.i)
    .map(({ e }) => e);
}

/** Show the card at all? Only when a night ran or a night head is blocked now. */
export function hasLastNight(data: LastNight | null | undefined): boolean {
  if (!data) return false;
  return (data.report?.entries.length ?? 0) > 0 || data.notices.length > 0;
}

/** "Answer" is offered while the live run still waits on its question. */
export function canAnswer(run: HeadRun | null | undefined): run is HeadRun {
  return !!run && run.state === "needs_you" && !!run.harness && !!run.runtime_slug;
}

export const NIGHT_CATEGORIES: NightCategory[] = ["passed", "failed", "needs_you", "blocked", "running"];

/** Reasons a pair may still be marked for tonight. */
const MARKABLE_BLOCKS = new Set(["engine_not_ready", "box_busy"]);

/** A pair that can be chosen for tonight (see module note). */
export function isMarkable(p: HeadPair): boolean {
  return p.startable || (p.status === "blocked" && MARKABLE_BLOCKS.has(p.reason_code ?? ""));
}

/**
 * Pairs for the "tonight" picker: today-only blocks become choosable. The
 * reason code stays, so the picker can still say "not running right now".
 */
export function pairsForTonight(pairs: HeadPair[]): HeadPair[] {
  return pairs.map((p) => (!p.startable && isMarkable(p) ? { ...p, startable: true } : p));
}

/** Initial pair for tonight: the launcher's rule (remembered local, else the
 *  local default) on the tonight-mapped list. Never a cloud pair. */
export function chooseTonightPair(resp: HeadPairsResponse | null | undefined, remembered: string | null): HeadPair | null {
  if (!resp) return null;
  const mapped = pairsForTonight(resp.pairs);
  const def = resp.default_pair ? (pairsForTonight([resp.default_pair])[0] ?? null) : null;
  return chooseInitialPair({ pairs: mapped, default_pair: def }, remembered);
}

/** i18n key under `nightShift.state` for an entry. A started entry shows its
 *  head state instead (`heads.state.*`). */
export function entryStateKey(entry: Pick<NightEntry, "state">): string {
  return `state.${entry.state}`;
}

/**
 * May this card be marked for tonight? Only when nobody else works on it: an
 * inbox card nothing holds (the mark will hold it), or an open card on
 * `manual_hold`. Same rule as the backend (`night.markable`, 409 `task_busy`).
 */
export function canMarkTonight(task: { status: string; run_control?: string | null }): boolean {
  if (task.status === "done" || task.status === "aborted") return false;
  const rc = task.run_control ?? null;
  return rc === "manual_hold" || (task.status === "inbox" && rc === null);
}

const REASONS = new Set([
  "task_moved",
  "task_move_failed",
  "lane_busy",
  "cloud_share",
  "engine_not_ready",
  "box_busy",
  "pair_blocked",
  "pair_gone",
  "repo_required",
  "head_active",
  "task_not_found",
  "spool_unavailable",
  "not_reached",
  "run_missing",
]);

/** i18n key under `nightShift.reason` for a waiting/skipped reason code. */
export function nightReasonKey(code: string | null | undefined): string {
  return code && REASONS.has(code) ? `reason.${code}` : "reason.other";
}

const ERROR_CODES = new Set([
  "heads_disabled",
  "task_not_found",
  "task_finished",
  "repo_required",
  "night_started",
  "night_busy",
  "task_busy",
  "head_active",
  "pair_blocked",
  "invalid_config",
]);

/** i18n key under `nightShift.errors` for any error of a night-shift call. */
export function nightErrorKey(err: unknown): string {
  const { code } = parseHeadError(err);
  return code && ERROR_CODES.has(code) ? `errors.${code}` : "errors.unknown";
}

/** Field codes of a 422 invalid_config. */
export function invalidConfigFields(err: unknown): string[] {
  const { code, detail } = parseHeadError(err);
  if (code !== "invalid_config" || !Array.isArray(detail.fields)) return [];
  return (detail.fields as unknown[]).map(String);
}

/** Counts per category of a morning report. */
export function reportCounts(report: Pick<NightReport, "entries"> | null | undefined): Record<NightCategory, number> {
  const out: Record<NightCategory, number> = { passed: 0, failed: 0, needs_you: 0, blocked: 0, running: 0 };
  for (const e of report?.entries ?? []) {
    if (e.category in out) out[e.category] += 1;
  }
  return out;
}

/** "HH:MM" check, same rule as the backend. */
export function isHHMM(value: string): boolean {
  return /^([01][0-9]|2[0-3]):[0-5][0-9]$/.test(value);
}

/** The browser's IANA zone (the Settings page offers it as a one-click value). */
export function browserTimeZone(): string | null {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone || null;
  } catch {
    return null;
  }
}

File: backend/app/routers/night_shift.py
"""Night shift API (ROADMAP E2) — mark tasks "run tonight", list tonight, settings.

  GET    /api/v1/night-shift/config            viewer    window settings + the next window
  PUT    /api/v1/night-shift/config            admin     change window / zone / cloud share / on-off
  GET    /api/v1/night-shift/tonight           viewer    marked tasks + their state, last report
  GET    /api/v1/night-shift/tasks/{task_id}   viewer    the mark of one task (or null)
  PUT    /api/v1/night-shift/tasks/{task_id}   operator  mark / change the pair
  DELETE /api/v1/night-shift/tasks/{task_id}   operator  unmark (before it started)

The start itself happens in mc-worker (``services/heads/night_shift.py``)
through the head launcher's start path. Errors carry a ``code`` for i18n.
Behind ``settings.heads_enabled`` like the head launcher.
"""
from __future__ import annotations

import time
import uuid
from dataclasses import asdict
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession

from app.auth import Role, require_role
from app.config import settings
from app.database import get_session
from app.models.task import Task
from app.routers.heads import run_view
from app.services.heads import box_guard, files, night, night_store, pairs
from app.services.heads.night_store import HOLD_REASON, NightMark
from app.services.heads.start import active_run_for_task

router = APIRouter(prefix="/api/v1/night-shift", tags=["night-shift"])

#: A pair that cannot start right now for these reasons may still be marked —
#: the engine can be up and the box free by tonight.
MARKABLE_BLOCKS = frozenset({"engine_not_ready", "box_busy"})
FINISHED = frozenset({"done", "aborted"})


def _enabled() -> None:
    if not settings.heads_enabled:
        raise HTTPException(status_code=404, detail={"code": "heads_disabled"})


def _err(status: int, code: str, **extra) -> HTTPException:
    return HTTPException(status_code=status, detail={"code": code, **extra})


def _iso(ts: float | None) -> str | None:
    return datetime.fromtimestamp(ts, UTC).isoformat() if ts else None


async def _config_view(session: AsyncSession) -> dict:
    cfg = await night_store.load_config(session)
    now = datetime.now(UTC)
    current = night.current_window(now, cfg)
    return {
        **asdict(cfg),
        "active": current is not None,
        "window": night.next_window(now, cfg).to_dict(),
    }


def mark_view(mark: NightMark, task: Task, runs: dict, now: float) -> dict:
    run = runs.get(mark.run_id) if mark.run_id else None
    if mark.run_id:
        state = "started"
    elif mark.gave_up:
        state = "skipped"
    elif mark.last_error:
        state = "waiting"
    else:
        state = "queued"
    return {
        "task_id": mark.task_id,
        "title": task.title,
        "task_status": task.status,
        "harness": mark.harness,
        "runtime_slug": mark.runtime_slug,
        "locality": mark.locality,
        "marked_at": _iso(mark.marked_at),
        "night": mark.night,
        "state": state,
        "reason": mark.gave_up or mark.last_error,
        "run": run_view(run, now) if run is not None else None,
    }


# ── Settings ────────────────────────────────────────────────────────────────


@router.get("/config", dependencies=[Depends(require_role(Role.VIEWER))])
async def get_config(session: AsyncSession = Depends(get_session)):
    _enabled()
    return await _config_view(session)


class ConfigBody(BaseModel):
    enabled: bool | None = None
    start: str | None = Field(default=None, max_length=5)
    end: str | None = Field(default=None, max_length=5)
    timezone: str | None = Field(default=None, max_length=64)
    cloud_share: int | None = None


@router.put("/config", dependencies=[Depends(require_role(Role.ADMIN))])
async def put_config(body: ConfigBody, session: AsyncSession = Depends(get_session)):
    _enabled()
    updates = body.model_dump(exclude_none=True)
    try:
        await night_store.save_config(session, updates)
    except ValueError as exc:
        fields = exc.args[0] if exc.args and isinstance(exc.args[0], list) else []
        raise _err(422, "invalid_config", fields=fields)
    return await _config_view(session)


# ── Tonight ────────────────────────────────────────────────────────────────


@router.get("/tonight", dependencies=[Depends(require_role(Role.VIEWER))])
async def get_tonight(session: AsyncSession = Depends(get_session)):
    _enabled()
    marks = night_store.list_marks()
    tasks: dict[str, Task] = {}
    if marks:
        ids = [uuid.UUID(m.task_id) for m in marks]
        tasks = {str(t.id): t for t in (await session.exec(select(Task).where(Task.id.in_(ids)))).all()}
    now = time.time()
    runs = {r.run_id: r for r in files.list_runs()} if any(m.run_id for m in marks) else {}
    entries = [mark_view(m, tasks[m.task_id], runs, now) for m in marks if m.task_id in tasks]
    report = night_store.latest_report()
    if report:
        report = {k: report.get(k) for k in ("night", "sent_at", "delivered", "entries")}
    return {"config": await _config_view(session), "entries": entries, "last_report": report}


# ── One task ───────────────────────────────────────────────────────────────


async def _task(session: AsyncSession, task_id: uuid.UUID) -> Task:
    task = await session.get(Task, task_id)
    if task is None:
        raise _err(404, "task_not_found")
    return task


@router.get("/tasks/{task_id}", dependencies=[Depends(require_role(Role.VIEWER))])
async def get_task_mark(task_id: uuid.UUID, session: AsyncSession = Depends(get_session)):
    _enabled()
    task = await _task(session, task_id)
    mark = night_store.load_mark(str(task_id))
    if mark is None:
        return {"mark": None}
    runs = {r.run_id: r for r in files.list_runs()} if mark.run_id else {}
    return {"mark": mark_view(mark, task, runs, time.time())}


class MarkBody(BaseModel):
    harness: str = Field(max_length=32)
    runtime_slug: str = Field(max_length=64)


@router.put("/tasks/{task_id}")
async def put_task_mark(
    task_id: uuid.UUID,
    body: MarkBody,
    session: AsyncSession = Depends(get_session),
    user=Depends(require_role(Role.OPERATOR)),
):
    _enabled()
    task = await _task(session, task_id)
    if task.status in FINISHED:
        raise _err(409, "task_finished")
    if not task.repo_id:
        raise _err(422, "repo_required")
    existing = night_store.load_mark(str(task_id))
    if existing is not None and existing.run_id:
        raise _err(409, "night_started")
    if active_run_for_task(str(task_id), time.time()) is not None:
        raise _err(409, "head_active")

    listing = await pairs.list_pairs(session, box_guard.occupancy())
    pair = next((p for p in listing["pairs"]
                 if p["harness"] == body.harness and p["runtime_slug"] == body.runtime_slug), None)
    if pair is None:
        raise _err(422, "pair_blocked", reason_code="runtime_not_offered")
    if pair["status"] == "blocked" and pair["reason_code"] not in MARKABLE_BLOCKS:
        raise _err(422, "pair_blocked", reason_code=pair["reason_code"])

    if existing is not None:
        mark = existing
        mark.harness, mark.runtime_slug, mark.locality = body.harness, body.runtime_slug, pair["locality"]
        mark.last_error, mark.gave_up = None, None
    else:
        uid = getattr(user, "id", None)
        mark = NightMark(task_id=str(task_id), harness=body.harness, runtime_slug=body.runtime_slug,
                    locality=pair["locality"], marked_at=time.time(), marked_by=str(uid) if uid else None)
    # An inbox card waits for the night: hold it so nothing else picks it up.
    if task.status == "inbox" and task.run_control is None:
        task.run_control = "manual_hold"
        task.hold_reason = HOLD_REASON
        mark.held = True
        session.add(task)
        await session.commit()
    night_store.save_mark(mark)
    return {"mark": mark_view(mark, task, {}, time.time())}


@router.delete("/tasks/{task_id}", dependencies=[Depends(require_role(Role.OPERATOR))])
async def delete_task_mark(task_id: uuid.UUID, session: AsyncSession = Depends(get_session)):
    _enabled()
    mark = night_store.load_mark(str(task_id))
    if mark is None:
        return {"mark": None}
    if mark.run_id:
        # The head is working (or done): stop it on the task; its outcome
        # still belongs into the morning report.
        raise _err(409, "night_started")
    task = await session.get(Task, task_id)
    if (
        task is not None and mark.held and task.status == "inbox"
        and task.run_control == "manual_hold" and task.hold_reason == HOLD_REASON
    ):
        task.run_control = None
        task.hold_reason = None
        session.add(task)
        await session.commit()
    night_store.delete_mark(str(task_id))
    return {"mark": None}
